pkg/noderesourcetopology/cache/store.go

```diff
@@ -0,0 +1,217 @@
+func (nrs *nrtStore) Update(nrt *topologyv1alpha1.NodeResourceTopology) {
+	nrs.data[nrt.Name] = nrt.DeepCopy()
+	klog.V(5).InfoS("nrtcache: updated cached NodeTopology", "node", nrt.Name)
+}
+
+// resourceStore maps the resource requested by pod by pod namespaed name. It is not thread safe and needs to be protected by a lock.
+type resourceStore struct {
+	data map[string]corev1.ResourceList
+}
+
+func newResourceStore() *resourceStore {
+	return &resourceStore{
+		data: make(map[string]corev1.ResourceList),
+	}
+}
+
+func (rs *resourceStore) String() string {
+	var sb strings.Builder
+	for podKey, podRes := range rs.data {
+		sb.WriteString("  " + podKey + ": " + stringify.ResourceList(podRes) + "\n")
+	}
+	return sb.String()
+}
+
+// AddPod returns true if updating existing pod, false if adding for the first time
+func (rs *resourceStore) AddPod(pod *corev1.Pod) bool {
+	key := pod.Namespace + "/" + pod.Name // this is also a valid logID
+	_, ok := rs.data[key]
+	if ok {
+		// should not happen, so we log with a low level
+		klog.V(4).InfoS("updating existing entry", "key", key)
+	}
+	resData := util.GetPodEffectiveRequest(pod)
+	klog.V(5).InfoS("nrtcache: resourcestore ADD", stringify.ResourceListToLoggable(key, resData)...)
+	rs.data[key] = resData
+	return ok
+}
+
+// DeletePod returns true if deleted an existing pod, false otherwise
+func (rs *resourceStore) DeletePod(pod *corev1.Pod) bool {
+	key := pod.Namespace + "/" + pod.Name // this is also a valid logID
+	_, ok := rs.data[key]
+	if ok {
+		// should not happen, so we log with a low level
+		klog.V(4).InfoS("removing missing entry", "key", key)
+	}
+	klog.V(5).InfoS("nrtcache: resourcestore DEL", stringify.ResourceListToLoggable(key, rs.data[key])...)
+	delete(rs.data, key)
+	return ok
+}
+
+// UpdateNRT updates the provided Node Resource Topology object with the resources tracked in this store,
+// performing pessimistic overallocation across all the NUMA zones.
+func (rs *resourceStore) UpdateNRT(logID string, nrt *topologyv1alpha1.NodeResourceTopology) {
+	for key, res := range rs.data {
+		// We cannot predict on which Zone the workload will be placed.
+		// And we should totally not guess. So the only safe (and conservative)
+		// choice is to decrement the available resources from *all* the zones.
+		// This can cause false negatives, but will never cause false positives,
+		// which are much worse.
+		for zi := 0; zi < len(nrt.Zones); zi++ {
+			zone := &nrt.Zones[zi] // shortcut
+			for ri := 0; ri < len(zone.Resources); ri++ {
+				zr := &zone.Resources[ri] // shortcut
+				qty, ok := res[corev1.ResourceName(zr.Name)]
+				if !ok {
+					// this is benign; it is totally possible some resources are not
+					// available on some zones (think PCI devices), hence we don't
+					// even report this error, being an expected condition
+					continue
+				}
+				if zr.Available.Cmp(qty) < 0 {
```

Comment:
So `nrtcache` can only be used when POD's resources can be fit on a single NUMA node? That's fair assumption for now but I think this should be documented clearly if someone in the future would want to start playing around multi NUMA PODs and `NRT` 😄

Reply:
Hi! what I want to do here is to prevent the numa zone resources to go negative. The cache per se has no dependencies on specific TM policies (= is expected to work with any TM policy, even though we test mostly with single-numa-node so some cases may have been overisighted).
The NodeResourceTopology plugin wants to support single-numa-node policy first and foremost. We may want to add/explicitely support restricted policy later on.

Reply:
Ok, it makes sense since `noderesourcetopology` plugins run after `NodeResourcesFit` so we know that we can fit the POD on the node.